### PR TITLE
Dev v3 refactor

### PR DESCRIPTION
**Pull request by BuddahLD**

### CODE REVIEW

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/HttpPureClient.java [hunk 1]
```diff
@@ -33,214 +33,312 @@
...
-    public PurekitProtosV3Storage.UserRecords getUsers(Collection<String> userIds) throws ProtocolHttpException, ProtocolException {
-        PurekitProtosV3Storage.GetUserRecords getUserRecords = PurekitProtosV3Storage.GetUserRecords.newBuilder().addAllUserIds(userIds).build();
+    public PurekitProtosV3Storage.UserRecords getUsers(Collection<String> userIds) 
+        throws ProtocolHttpException, ProtocolException {
 
-        return this.client.firePost(
+        if (userIds == null) {
+            throw new NullArgumentException("userIds");
+        }
+//        if (userIds.isEmpty()) { // FIXME can we pass empty userIds?
```
Comment (SanjoDeundiak) [inline purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/HttpPureClient.java]: Uncomment

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/HttpPureClient.java [hunk 2]
```diff
@@ -33,214 +33,312 @@
...
+     * @throws ProtocolException Thrown if an error from the PHE service has been parsed
+     * successfully.
+     * @throws ProtocolHttpException Thrown if an error from the PHE service has NOT been parsed
+     * successfully. Represents a regular HTTP exception with code and message.
      */
-    public void deleteUser(String userId, boolean cascade) throws ProtocolHttpException, ProtocolException {
-        // TODO: parameters ideally should not be added directly to url string
+    public void deleteUser(String userId, boolean cascade) 
+        throws ProtocolHttpException, ProtocolException {
+        // TODO: parameters ideally should not be added directly to url string -> explain please
```
Comment (SanjoDeundiak) [inline purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/HttpPureClient.java]: Parameters are passed as argument to firePost/firePut. How fireGet/fireDelete is different?

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/HttpPureClient.java [hunk 3]
```diff
@@ -33,214 +33,312 @@
...
         );
     }
+
+    /**
+     * Checks whether userId and dataId is null or empty.
+     *
+     * @param userId User Ids.
+     * @param dataId Data Id.
+     */
+    private void validateUserAndData(String userId, String dataId) {
```
Comment (SanjoDeundiak) [inline purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/HttpPureClient.java]: Make this method accept 1 arbitrary String and parameter name, then it can be used everywhere

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/Pure.java [hunk 4]
```diff
@@ -106,114 +126,65 @@ public Pure(PureContext context) throws CryptoException, PureException {
...
-
-        byte[] phek = client.checkResponseAndDecrypt(passwordHash, userRecord.getPheRecord(), response.getResponse().toByteArray());
-
-        if (phek.length == 0) {
-            throw new PureException(PureException.ErrorCode.INVALID_PASSWORD);
+        VirgilKeyPair ukp;
+        try {
+            ukp = this.crypto.importPrivateKey(uskData);
+        } catch (CryptoException exception) {
+            throw new IllegalStateException("Exception while importing key. This should not"
```
Comment (SanjoDeundiak) [inline purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/Pure.java]: uskData depends on what is in the storage. Any error that appears while working with external systems or their data MUST NOT be RuntimeException

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/Pure.java [hunk 5]
```diff
@@ -225,209 +196,250 @@ public AuthResult authenticateUser(String userId, String password, String sessio
...
-    public PureGrant decryptGrantFromUser(String encryptedGrantString) throws Exception {
-        if (encryptedGrantString == null || encryptedGrantString.isEmpty()) {
-            throw new NullPointerException();
-        }
+    public PureGrant decryptGrantFromUser(String encryptedGrantString)
+        throws ProtocolException, ProtocolHttpException, PureException,
+        InvalidProtocolBufferException {
 
-        byte[] encryptedGrantData = Base64.getDecoder().decode(encryptedGrantString);
+        ValidateUtils.checkNullOrEmpty(new HashMap<String, String>() {{
```
Comment (SanjoDeundiak) [inline purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/Pure.java]: This is overkill

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/PureStorage.java [hunk 6]
```diff
@@ -33,95 +33,172 @@
...
+     *
+     * @param userRecord User record.
+     *
+     * @throws ProtocolException Thrown if an error from the PHE service has been parsed
+     * successfully.
+     * @throws ProtocolHttpException Thrown if an error from the PHE service has NOT been parsed
+     * successfully. Represents a regular HTTP exception with code and message.
      */
-    void insertUser(UserRecord userRecord) throws Exception;
+    void insertUser(UserRecord userRecord)
```
Comment (SanjoDeundiak) [inline purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/PureStorage.java]: Why interface methods throw exceptions from concrete implementation?

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/PureStorage.java [hunk 7]
```diff
@@ -33,95 +33,172 @@
...
-     * Calling method will request records until empty value is returned
-     * @param pheRecordVersion PheRecordVersion
-     * @return UserRecords
-     * @throws Exception FIXME
+     * This method throws MethodNotImplementedException, as in case of using Virgil Cloud storage,
+     * rotation happens on Virgil side.
+     *
+     * @param pheRecordVersion PHE record version.
+     *
+     * @return always throws NotImplementedException.
```
Comment (SanjoDeundiak) [inline purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/PureStorage.java]: How could you know what implementation throws?

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/VirgilCloudPureStorage.java [hunk 8]
```diff
@@ -210,156 +167,244 @@ public UserRecord selectUser(String userId) throws Exception {
...
+        boolean verified;
+        try {
+            verified = crypto.verifySignature(
+                protobufRecord.getSignature().toByteArray(),
                 protobufRecord.getCellKeySigned().toByteArray(),
-                this.signingKey.getPublicKey());
+                this.signingKey.getPublicKey()
+            );
+        } catch (VerificationException exception) {
+            throw new IllegalStateException("This should not happen. Please, contact developer.",
```
Comment (SanjoDeundiak) [inline purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/VirgilCloudPureStorage.java]: When does it happen?

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/VirgilCloudPureStorage.java [hunk 9]
```diff
@@ -210,156 +167,244 @@ public UserRecord selectUser(String userId) throws Exception {
...
         }
 
-        PurekitProtosV3Storage.CellKeySigned r = PurekitProtosV3Storage.CellKeySigned.parseFrom(protobufRecord.getCellKeySigned());
+        PurekitProtosV3Storage.CellKeySigned keySigned;
+        try {
+            keySigned =
+                PurekitProtosV3Storage.CellKeySigned.parseFrom(protobufRecord.getCellKeySigned());
+        } catch (InvalidProtocolBufferException e) {
+            throw new IllegalStateException("Error while parsing protobuf message from "
+                                                + "PHE service.");
```
Comment (SanjoDeundiak) [inline purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/VirgilCloudPureStorage.java]: Pure

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/exception/MethodNotImplementedException.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.virgilsecurity.purekit.pure.exception;
+
+/**
+ * MethodNotImplementedException class.
+ */
+public class MethodNotImplementedException extends RuntimeException {
```
Comment (SanjoDeundiak) [inline purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/exception/MethodNotImplementedException.java]: UnsupportedOperationException?

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/exception/ServiceErrorCode.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+package com.virgilsecurity.purekit.pure.exception;
+
+/**
+ * PHE service error codes.
+ */
+public enum ServiceErrorCode {
+    USER_NOT_FOUND(50003),
+    CELL_KEY_NOT_FOUND(50004),
+    CELL_KEY_ALREADY_EXISTS(50006); // FIXME do we need error code 0 ?
```
Comment (SanjoDeundiak) [inline purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/exception/ServiceErrorCode.java]: Why don't we?

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/utils/ValidateUtils.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * ValidateUtils class.
+ */
+public class ValidateUtils {
+
+  /**
+   * Throws NullArgumentException if an argument is null and EmptyArgumentException if argument
+   * is empty.
+   */
+  public static void checkNullOrEmpty(Map<String, String> argumentsToCheck) {
```
Comment (SanjoDeundiak) [inline purekit/src/main/kotlin/com/virgilsecurity/purekit/utils/ValidateUtils.java]: overkill

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/PureStorage.java [hunk 13]
```diff
@@ -33,95 +33,172 @@
...
+     *
+     * @throws PureException If user Id duplicate has been found or user id mismatches the one
+     * from a server. Use {@link PureException#getErrorStatus()} to know the particular case.
+     * @throws ProtocolException Thrown if an error from the PHE service has been parsed
+     * successfully.
+     * @throws ProtocolHttpException Thrown if an error from the PHE service has NOT been parsed
+     * successfully. Represents a regular HTTP exception with code and message.
      */
-    Iterable<UserRecord> selectUsers(Collection<String> userIds) throws Exception;
+    Collection<UserRecord> selectUsers(Collection<String> userIds)
```
Comment (SanjoDeundiak) [inline purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/PureStorage.java]: What's the reason of this change?

File: purekit/src/test/java/com/virgilsecurity/purekit/protocol/PureTestJava.java [hunk 14]
```diff
@@ -824,15 +829,15 @@ void delete_key__new_key__should_delete(String pheServerAddress,
...
+                assertEquals(PureException.ErrorStatus.CELL_KEY_NOT_FOUND_IN_STORAGE, e.getErrorStatus());
             }
         }
         catch (Exception | ProtocolHttpException e) {
             fail(e);
         }
     }
 
-    // TODO: Test hashes encryption
+    // TODO: Test hashes encryption -> ????
```
Comment (SanjoDeundiak) [inline purekit/src/test/java/com/virgilsecurity/purekit/protocol/PureTestJava.java]: Add test that proves that password hashed and encrypted, stored and can be decrypted

**Review by SanjoDeundiak — CHANGES_REQUESTED** [review]

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/HttpPureClient.java (hunk 2 above)
Comment (BavovnaIsTaken) [reply purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/HttpPureClient.java]: POST has body, while GET is just a call to endpoint with query in url?

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/Pure.java (hunk 5 above)
Comment (BavovnaIsTaken) [reply purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/Pure.java]: just similar approach as in all other methods, so all checks are the same

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/PureStorage.java (hunk 13 above)
Comment (BavovnaIsTaken) [reply purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/PureStorage.java]: Map is Iterable as well. Are we ok with passing Map here, for example?

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/HttpPureClient.java (hunk 2 above)
Comment (SanjoDeundiak) [reply purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/HttpPureClient.java]: POST/PUT encode parameters to body, GET/DELETE encode parameters to url, but the idea is the same. I don't understand why in case of POST and PUT we pass parameters to firePost/firePut and in case of GET and DELETE we should handle it ourselves before calling fireGet/fireDelete

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/Pure.java (hunk 5 above)
Comment (SanjoDeundiak) [reply purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/Pure.java]: Creating HashMap on every function call just to check arguments nullability is overkill, slow and probably could not be optimised by compiler.

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/PureStorage.java (hunk 13 above)
Comment (SanjoDeundiak) [reply purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/PureStorage.java]: Interface should be as less restrictive as possible. AFAIK usages of this interface don't need anything Iterable doesn't have

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/exception/UnsupportedOperationException.java [hunk 15]
```diff
@@ -34,26 +34,26 @@
 package com.virgilsecurity.purekit.pure.exception;
 
 /**
- * MethodNotImplementedException class.
+ * UnsupportedOperationException class.
  */
-public class MethodNotImplementedException extends RuntimeException {
+public class UnsupportedOperationException extends RuntimeException {
```
Comment (SanjoDeundiak) [inline purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/exception/UnsupportedOperationException.java]: https://docs.oracle.com/javase/7/docs/api/java/lang/UnsupportedOperationException.html

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/VirgilCloudPureStorage.java (hunk 9 above)
Comment (BavovnaIsTaken) [reply purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/VirgilCloudPureStorage.java]: removed IllegalStateException at all

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/VirgilCloudPureStorage.java (hunk 8 above)
Comment (BavovnaIsTaken) [reply purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/VirgilCloudPureStorage.java]: removed IllegalStateException at all

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/VirgilCloudPureStorage.java [hunk 16]
```diff
@@ -125,9 +173,27 @@ public UserRecord selectUser(String userId)
...
+     * @throws ProtocolHttpException Thrown if an error from the PHE service has NOT been parsed
+     * successfully. Represents a regular HTTP exception with code and message.
+     * @throws InvalidProtocolBufferException If a PurekitProtosV3Storage.UserRecord received from
+     * a server cannot be parsed as a Protobuf message.
+     * @throws VerificationException If signature verification operation failed.
+     */
     @Override
-    public Iterable<UserRecord> selectUsers(Collection<String> userIds) // FIXME maybe just use Set instead of Iterable? so we can avoid DUPLICATE_USER_ID
-        throws PureException, ProtocolException, ProtocolHttpException {
+    public Iterable<UserRecord> selectUsers(Collection<String> userIds) // FIXME maybe just use Set instead of Collection? so we can avoid DUPLICATE_USER_ID
```
Comment (SanjoDeundiak) [inline purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/VirgilCloudPureStorage.java]: Remove fix

**Review by SanjoDeundiak — CHANGES_REQUESTED** [review]

File: purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/VirgilCloudPureStorage.java (hunk 16 above)
Comment (BavovnaIsTaken) [reply purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/VirgilCloudPureStorage.java]: what's about Set? the question itself

Comment (SanjoDeundiak) [reply purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/VirgilCloudPureStorage.java]: Do you now storage that returns collection of entities in form of Set? I don't. And if there is no such storages, fetching all entities and then pushing them to Set is overhead

Comment (BavovnaIsTaken) [reply purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/VirgilCloudPureStorage.java]: Collection<String> userIds -> Set<String>
it's an argument, not a return type

Comment (SanjoDeundiak) [reply purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/VirgilCloudPureStorage.java]: I though we're talking about return type. Argument may be changed to Set, that doesn't matter so much

**Review by SanjoDeundiak — APPROVED** [review]
